src/mscorlib/shared/System/Globalization/TextInfo.cs

```diff
@@ -201,6 +202,223 @@ public unsafe virtual string ToLower(string str)
             return ChangeCase(str, toUpper: false);
         }
 
+        private unsafe char ChangeCase(char c, bool toUpper)
+        {
+            Debug.Assert(!_invariantMode);
+            
+            char dst = default;
+            ChangeCase(&c, 1, &dst, 1, toUpper);
+            return dst;
+        }
+
+        private unsafe string ChangeCase(string source, bool toUpper)
+        {
+            Debug.Assert(!_invariantMode);
+            Debug.Assert(source != null);
+
+            // If the string is empty, we're done.
+            if (source.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            int sourcePos = 0;
+            string result = null;
+
+            // If this culture's casing for ASCII is the same as invariant, try to take
+            // a fast path that'll work in managed code and ASCII rather than calling out
+            // to the OS for culture-aware casing.
+            if (IsAsciiCasingSameAsInvariant)
+            {
+                if (toUpper)
+                {
+                    // Loop through each character.
+                    for (sourcePos = 0; sourcePos < source.Length; sourcePos++)
+                    {
+                        // If the character is lower-case, we're going to need to allocate a string.
+                        char c = source[sourcePos];
```

Comment:
>char c = source[sourcePos]; [](start = 24, length = 27)

Is it better to use a pointer to the string instead of using the string directly here? I think this can avoid the bound checking when accessing the indexed characters.

Reply:
The JIT should be able to avoid the bounds checking here, as it can see that sourcePos is always within the bounds of the string based on the loop construction.  I'd tried using pointers, but it didn't help with throughput and it made the code more unsafe, so I reverted it.